### PR TITLE
First cut of the GoPiGo3 driver, lacking grove peripherals

### PR DESCRIPTION
**Pull request by ulisesflynn**
This builds on the collaboration with @brimstone and @iplay88keys from the GopherCon 2017 hack day. It makes use of the golang.org/x/exp/io/spi library for SPI support, and is based from GoPiGo3 python library. 

It's fully functioning, minus the grove peripherals, but be sure to run the GoPiGo3 shell script: https://www.dexterindustries.com/GoPiGo/get-started-with-the-gopigo3-raspberry-pi-robot/3-program-your-raspberry-pi-robot/python-programming-language/ before running this on the raspberry pi.

### CODE REVIEW

**Comment by deadprogram** [comment]
Hi @ulisesflynn this looks really awesome! Could you please change it to be against the `dev` branch instead of `master`? Thank you!

**Comment by deadprogram** [comment]
Thanks for changing the branch for merge, @ulisesflynn 

Last request I think that the file `drivers/spi/gopigo3_driver.go` should go into `platforms/gopigo/gopigo3_driver.go` instead since it is specific to the GoPiGo "platform", similar to the PR https://github.com/hybridgroup/gobot/pull/438 that @iplay88keys submitted. 

When we have an ssd1306 driver that supports the SPI interface, that would be an example of something that would go into `drivers\spi\ssd1306_driver.go`.

This PR has general purpose SPI communication on the raspi, as well as the GoPiGo implementation, so I think it is a great push in that direction. I'd love to have the platform move I mention above, what  does everyone think?

**Comment by ulisesflynn** [comment]
Thanks Ron for the feedback. I did see putting the gopigo3 into a platform from the PR from @iplay88keys, but I was not sure at the which way to go.  However it does make sense to move it to a platform, and if everyone is good with it, I'll make the change.

File: examples/gopigo3.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"fmt"
+	"time"
+
+	"gobot.io/x/gobot"
+	g "gobot.io/x/gobot/platforms/gopigo3"
+)
+
+func main() {
+	gopigo3Adaptor := g.NewAdaptor("/dev/spidev0.1", 0, 500000)
```
Comment (deadprogram) [inline examples/gopigo3.go]: This example will need to be updated, right @ulisesflynn ?

File: platforms/gopigo3/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import (
+	"fmt"
+	"time"
+
+	"gobot.io/x/gobot"
+	g "gobot.io/x/gobot/platforms/gopigo3"
+)
+
+func main() {
+	gopigo3Adaptor := g.NewAdaptor("/dev/spidev0.1", 0, 500000)
```
Comment (deadprogram) [inline platforms/gopigo3/README.md]: Docs updated needed here, I think.

**Comment by deadprogram** [comment]
I'm getting very excited about this!

**Comment by deadprogram** [comment]
@ulisesflynn looks like this is ready! Anything else you want to add, or can I merge into `dev` now?

**Comment by ulisesflynn** [comment]
Thanks @deadprogram I think this is good for now, and I've tested this with the gopigo3 hardware.

**Comment by deadprogram** [comment]
I am now merging into `dev` branch. 

Thank you very much @ulisesflynn @brimstone and @iplay88keys for all of the code and efforts! This work both adds the GoPiGo3 platform, and also moves forward the overall goal of adding SPI interface to Gobot.

:1st_place_medal: :1st_place_medal: :1st_place_medal:

**Pull request merged**